enhancements/machine-api/spot-instances.md

```diff
@@ -0,0 +1,359 @@
+
+- The actuator is responsible for creation of the instance only and should not attempt to remediate problems
+
+- The actuator should not attempt to verify that an instance can be created before attempting to create the instance
+  - If the cloud provider does not have capacity, the Machine Health Checker can (given required MHC) remove the Machine after a period.
+    MachineSet will ensure the correct number of Machines are created.
+
+- There’s a [2 hours time frame](https://github.com/openshift/cluster-machine-approver/blob/master/csr_check.go#L38) in OpenShift for a machine to become a node.
+  If the Spot request is not satisfied during this time frame, it is assumed that the instance will never become a node.
+  This machine would be remediated by a given MHC.
+
+
+### Implementation Details
+
+#### Cloud Provider Implementation Specifics
+
+##### AWS
+
+###### Launching instances
+
+To launch an instance as a Spot instance on AWS, a [SpotMarketOptions](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#SpotMarketOptions)
+needs to be added to the `RunInstancesInput`. Within this there are 3 options that matter:
+
+- InstanceInterruptionBehaviour (default: terminate): This must be set to `terminate` otherwise the SpotInstanceType cannot be `one-time`
+
+- SpotInstanceType (default: one-time): This must be set to `one-time` to ensure that each Machine only creates on EC2 instance and that the spot request is
+
+- MaxPrice (default: On-Demand price): This can be **optionally** set to a string representation of the hourly maximum spot price.
+  If not set, the option will default to the On-Demand price of the EC2 instance type requested
+
+The only option from this that needs exposing to the user from this is the `MaxPrice`, this option should be in an optional struct, if the struct is not nil,
+then spot instances should be used, if the MaxPrice is set, this should be used instead of the default On-Demand price.
+
+```
+type SpotMarketOptions struct {
+  MaxPrice *string `json:”maxPrice,omitempty”`
+}
+
+type AWSMachineProviderConfig struct {
+  ...
+
+  SpotMarketOptions *SpotMarketOptions `json:”spotMarketOptions,omitempty”`
+}
+```
+
+###### Termination Notices
+Termination notices on AWS are provided via the EC2 metadata service up to 2 minutes before an instance is due to be preempted.
+A DaemonSet should be deployed on the cluster which runs Pods on nodes labelled `machine.openshift.io/terminate-on-preemption: true`.
+Users of Spot instances could optionally add this label to their Machine’s spec to opt in to this feature.
+
+The DaemonSet itself would poll the EC2 metadata service for a termination notice.
+As soon as it receives an `OK` response (a non-terminated instance returns a `404` response), it would delete the Machine object associated with the node that it is running on.
+This deletion will trigger the Machine controller to start draining the node, ready for its termination.
```

Comment:
The deletion will also trigger the machineset to create a replacement.  This is most likely going to not work well if we're losing an instance due to capacity claw-back from the provider.  This may be exactly what the user wants, a new spot request is generated, and fulfilled when capacity becomes available, or it may not.

We need to think about the UX of scaling up or down a machineset with spot instances in various states.

Reply:
> The deletion will also trigger the machineset to create a replacement.

Assuming the Machine is owned by a MachineSet, I think the implication of running a MachineSet is that I want x running machines at any one time, so if 1 of these is terminated, a new one should be brought up in it's place. 
I don't think deleting the machine breaks the UX of a MachineSet in that respect.

Where it may fall down is, as you say, if there is no capacity, at this point the MachineSet is unable to fulfil its guarantee, but I think if people are using spot instances, this is something they should be aware of and should configure other systems to take up any potential slack (cluster autoscaler I'm thinking).

> We need to think about the UX of scaling up or down a machineset with spot instances in various states.

I haven't looked into the logic of how a MachineSet determines which of its nodes to terminate if there is a scaling event, that's a good point, we may need to add some extra logic to it depending on how these things play out. I'll look into this a bit more.

Reply:
I had a look into this and, if I understand the logic correctly, it looks like any node which is already terminating or has not yet fully started, will be removed from the MachineSet first. So I believe the UX is safe. If you scale up, can't get new nodes, then scale back down, those failed new nodes would be the ones to be removed, and not the healthy ones that have your workload on